### PR TITLE
Fix #1146 - dispatch_sync to the main queue when manipulating UIApplication. applicationIconBadgeNumber

### LINKED ISSUE #1146: XCode 9: Main Thread Checker: UI API called from background thread: -[UIApplication applicationIconBadgeNumber] must be called from main thread only
Using the 1.14.4 version with XCode 9 on iOS 11 with the new main thread sanitizer activated.

This is the backtrace I got:

```
Main Thread Checker: UI API called on a background thread: -[UIApplication applicationIconBadgeNumber]
PID: 808, TID: 954593, Thread name: (none), Queue name: com.apple.root.default-qos, QoS: 21
Backtrace:
4   PetMonitor                          0x0000000103fb5df4 -[PFApplication iconBadgeNumber] + 48
5   PetMonitor                          0x0000000103fb1164 -[PFInstallation _updateBadgeFromDevice] + 80
6   PetMonitor                          0x0000000103fb1020 -[PFInstallation _objectWillSave] + 96
7   PetMonitor                          0x0000000103f57728 __56-[PFObject(Private) _enqueueSaveEventuallyWithChildren:]_block_invoke.430 + 84
8   PetMonitor                          0x0000000103f27a00 __62-[BFTask continueWithExecutor:successBlock:cancellationToken:]_block_invoke + 92
9   PetMonitor                          0x0000000103f272a4 __55-[BFTask continueWithExecutor:block:cancellationToken:]_block_invoke + 100
10  PetMonitor                          0x0000000103f285f8 __29+[BFExecutor defaultExecutor]_block_invoke_2 + 128
11  PetMonitor                          0x0000000103f28b58 -[BFExecutor execute:] + 76
12  PetMonitor                          0x0000000103f26e88 -[BFTask runContinuations] + 356
13  PetMonitor                          0x0000000103f26880 -[BFTask trySetResult:] + 152
14  PetMonitor                          0x0000000103f249cc -[BFTaskCompletionSource setResult:] + 80
15  PetMonitor                          0x0000000103f27580 __55-[BFTask continueWithExecutor:block:cancellationToken:]_block_invoke_2 + 300
16  PetMonitor                          0x0000000103f27364 __55-[BFTask continueWithExecutor:block:cancellationToken:]_block_invoke + 292
17  libdispatch.dylib                   0x000000010685d654 _dispatch_call_block_and_release + 24
18  libdispatch.dylib                   0x000000010685d614 _dispatch_client_callout + 16
19  libdispatch.dylib                   0x0000000106869f00 _dispatch_queue_override_invoke + 1112
20  libdispatch.dylib                   0x000000010686fe4c _dispatch_root_queue_drain + 596
21  libdispatch.dylib                   0x000000010686fb98 _dispatch_worker_thread4 + 64
22  libsystem_pthread.dylib             0x00000001a8f71338 _pthread_wqthread + 1260
23  libsystem_pthread.dylib             0x00000001a8f70e40 start_wqthread + 4
```

Looks like ``PFApplication`` needs to dispatch back to the main queue to interact with ``UIApplication``.

### PR DESCRIPTION
**Pull request by mman**
Make sure that we are getting and setting ``UIApplication. applicationIconBadgeNumber`` from the main thread.

### CODE REVIEW

File: Parse/Internal/PFApplication.m [hunk 1]
```diff
@@ -50,7 +50,9 @@ - (NSInteger)iconBadgeNumber {
 #if TARGET_OS_WATCH || TARGET_OS_TV
     return 0;
 #elif TARGET_OS_IOS
-    return self.systemApplication.applicationIconBadgeNumber;
+    __block NSInteger returnValue = 0;
+    dispatch_sync(dispatch_get_main_queue(), ^{ returnValue = self.systemApplication.applicationIconBadgeNumber; });
```
Comment (flovilmart) [inline Parse/Internal/PFApplication.m]: if called on main, this will deadlock!

**Comment by mman** [comment]
Updated to not deadlock when called from the main thread. Implemented the same way ``PFLocationManager`` uses dispatch_sync to the main queue.

**Comment by mman** [comment]
I've updated this pull request to base it off the latest master, the changes look good but the commit history is somewhat garbled... not that of a big friend with git... please recommend what is needed to cleanup this

**Comment by flovilmart** [comment]
I'm not sure what's going on with the iOS tests, did you run `bundle exec rake test:ios` ?

**Comment by mman** [comment]
Tried once again now, tests passing fine on my box

**Comment by mman** [comment]
Thanks @flovimart, I'm OOO for several more days so can't spend any time on it until next week.

**Comment by funnel20** [comment]
Can this fix in another project help?
https://github.com/urbanairship/ios-library/issues/139

**Comment by flovilmart** [comment]
The issue is actually a main thread deadlock now, and rhat’s The last thing we want.

**Comment by flovilmart** [comment]
@mman it's gonna be superseded by #1185

**Comment by mman** [comment]
Thanks @flovilmart , excellent work!

**Comment by wdcurry** [comment]
I am still seeing this issue, with Parse v1.15.3. I am noting that ParseUI is behind in its' Bolts refs. 

Is there a standard way to bring everything uptodate to avoid these troublesome background issues?

**Comment by flovilmart** [comment]
@wdcurry, managing and maintaining dependencies is an art, and it takes time :)

We need to work on the ParseUI framework and bump the dependencies there. Is it something you want to explore?

**Comment by wdcurry** [comment]
If i had the time and expertise i would love to! At this stage, i took on Parse Server a while back to free me up to deliver client-requested apps into iOS and have my hands full at that end of the spectrum. I am simply trying to update an app that is needing tweaks for iOS11, and am noting these new warnings relating to background threads.

**Comment by flovilmart** [comment]
Yeah, probalbly, is your parse-server instance running on HTTPS?

**Comment by wdcurry** [comment]
Yes it is. I just ran a quick check for "parse.com" and found at least one default variable pointing back there. I wonder if somehow this is still happening? I will go back a rake through my ATS efforts to try and track this down.

**Comment by wdcurry** [comment]
It's puzzling. I am getting this error repeatedly now, but the app seems to be functioning, and the exception breakpoint is NOT being hit:
`2017-11-25 10:17:25.793333-0400 {myApp}[1794:976580] Task <6F0BDF76-4AA5-432A-B9E8-A61D95E1BABA>.<19> HTTP load failed (error code: -1003 [12:8])
2017-11-25 10:17:25.799795-0400 {myApp}[1794:976569] Task <6F0BDF76-4AA5-432A-B9E8-A61D95E1BABA>.<19> finished with error - code: -1003
2017-11-25 10:17:25.801680-0400 {myApp}[1794:976580] [Error]: A server with the specified hostname could not be found. (Code: 100, Version: 1.15.3)
2017-11-25 10:17:25.802156-0400 {myApp}[1794:976580] [Error]: Network connection failed. Making attempt 3 after sleeping for 5.185565 seconds.`

**Comment by flovilmart** [comment]
Is your parse-server URL correct? Set correctly in the configuration etc?

**Comment by wdcurry** [comment]
Yes, that’s the oddity here. The app has been running fine for quite a while. This has only been appearing since the upgrades to High Sierra and/or ios11.. the app runs fine in dev mode, just that these are appearing in debug window but not triggering my exception breakpoint. 

Odd.

**Comment by wdcurry** [comment]
For completion: i tracked the issue down. I simply do not like that Bolts seems to bury the error (tis my guess as i have seen this before) and keeps the exception from firing but passed out a generic message without details. 

The issue was simple once my mind cleared: i had a few older records that were carried from Parse.com's era, and that featured PFFiles stored there. That domain is dead of course. Once i walked each record involved, resetting each avatar file, the error cleared completely.

So simple once you see the issue. ;)

**Comment by flovilmart** [comment]
Did you not transfer your files?

**Comment by wdcurry** [comment]
yes, but the transfer process doesn't (at least in my case) changed the stored references within the PFFile record within my custom data objects.

**Pull request closed without merging**